### ISSUE TITLE
12.5.1 breaks LWRP code that works with <= 12.4.3/4

### ISSUE THREAD

**Issue by Sauraus** [chef/chef#4124]
The following LWRP code: https://gist.github.com/Sauraus/6b699f7726957bb8f577 works 100% in 12.4.3/4 and does absolutely nothing in 12.5.1.

The worst part about it is that in 12.5.1 there is no error thrown or anything chef-client just says (up-to-date) and continues on without actually executing the action method.

**Comment by lamont-granquist** [comment]
so, what action are you running?  `:install` presumably?  and on windows or mac?  and can you give `-l debug` output so that we can see what the provider resolver is doing to resolve the class that handles the action?

**Comment by Sauraus** [comment]
https://gist.github.com/Sauraus/1f666c0042988c9cb106 Shows the output of kitchen converge, first file is the output with chef-client 12.4.3 and the second is 12.5.1.

in BOTH cases exactly the same recipe and LWRP code is being used.

**Comment by lamont-granquist** [comment]
Well that's weird, I see the correct classes for the LWRP being picked in the debug output, and even see the remote_file and two execute resources being referenced, so it *looks* like your LWRP provider code was being run and was constructing a resource collection with those resources in it.

One thing to ask here is that you're doing a 'kitchen converge' rather than 'kitchen test' and is it possible that the prior run installed it and its really the uninstall action which is broken somehow?

If you 'kitchen test' and scrap the virt and start over does it run the install correctly?

**Comment by lamont-granquist** [comment]
So one thing is that here you're calling install_dir on your new_resource:

https://gist.github.com/Sauraus/6b699f7726957bb8f577#file-editor-rb-L38

You don't actually have an install_dir on your resource, you've got install_root, while the provider itself has an install_dir method.  Honestly not certain how that worked in 12.4.3 if I'm understanding the code correctly, but I'm suspecting that method_missing was doing some crazy magic that got removed...

**Comment by ranjib** [comment]
we are hitting several similar issues. i am changing codes on our side to use `action :install ` from `def action_install`

**Comment by lamont-granquist** [comment]
So that works because we inject the lexical scope of the enclosing provider into the sub-resources in LWRPs via method_missing.  That was done for https://tickets.opscode.com/browse/CHEF-614 in https://github.com/chef/chef/commit/8d97aadcea8e407a39190e3baf8df138cd69077d which is pretty much unchanged here:

https://github.com/chef/chef/blob/b6ba4d29184cfc14b1256fdb1f442a321047de84/lib/chef/resource.rb#L1288-L1297

I don't think that's it, but if you fixed that and it went away that'd be curious.

**Comment by lamont-granquist** [comment]
@ranjib if you were trying to use_inline_resource and write `def action_install` that was always broken because use_inline_resources always required calling the action DSL method in order to wire up the child run_context and sub-resource_collection.

**Comment by ranjib** [comment]
@lamont-granquist yeah.. but it was working earlier, hence i changed it on our side, i realized probably it was working due to some bug :-) . i didnt chase it any further. in our lwrp definition they were also predicated by guards, which were not being evaluated.

**Comment by lamont-granquist** [comment]
Yeah, I suspect that "working" is the wrong word to use there.

If you look at the "old" use_inline_resources code before all the recent changes, it looks like this:

https://github.com/chef/chef/blob/3a87159beb91315401a3b88ea847ae6821e0ab88/lib/chef/provider/lwrp_base.rb#L30-L70

All that does is monkeypatch your provider DSL so that when you type `action :install` and give it a block you don't just get a `def action_install` with the block, but it gets wrapped with the `recipe_eval_with_update_check` which is where all the sub-resource_collection magic is.

<lecture>
I'm tempted to file this under the category of everyone who has been writing "HWLRPs" or whatever we're calling them needing to take a deep breath and seriously reconsider what they've been doing -- you're all picking a fight with the LWRP DSL and you're kinda loosing.  There's also awful stuff around the way that class names get auto-wired up to the chef DSL and provider_resolver (which date back to ancient times) which troll people.

You should check that things like Unity::Resources::UnityEditor::Resource isn't getting wired up to 'resource' in the Chef DSL for you -- this created a problem when MySQL had a class structure with a MysqlClient base class with MysqlClient::RHEL and MysqlClient::Debian subclasses.  The dynamic resource/provider resolver was finding the base class when you typed `mysql_client` in the DSL which was causing chaos so the solution was to rename them MysqlClientBase, MysqlClientRHEL and MysqlClientDebian or something like that.
</lecture>

So, yeah if you can fix this by changing `def action_install` to `action :install` then I think this is not a bug in core chef (it may troll a *lot* of cookbooks, but they were all buggy to start with and people have been typing use_inline_resources and not getting that behavior at all...)

And if you'd like an opinion from a core chef developer with 4+ years of experience in all this codebase, I don't think the community is smart enough to understand the edge cases of abandoning the LWRP DSL the way that everyone has done--mostly because I'm learning I'm not that smart, and I'm the one seeing all the bugs.  The two reasons I know of for abandoning the LWRP DSL are that we thought we couldn't rename LWRPs (because we never connected the dots with the Chef 11.x version of Chef::Resource.provides) and there were at least two horrible issues with ChefSpec testing which have been fixed (the memory leak issues by constantly unloading and reloading classes which went away in Chef 12, and some recent work on the way that ChefSpec failed to correctly create sub-run_contexts for LWRPs).  I've always thought that the shift away from the LWRP smelled wrong, now I'm pretty much convinced that it is wrong.  It makes the problem much more complicated, and you get edge cases like this wrong, and there's now no benefit to it that we're aware of.

**Comment by Sauraus** [comment]
@lamont-granquist I updated my LWRP resource gist with the working code sorry about that.

As for kitchen converge vs. test, I run a kitchen destroy between the two runs and modify .kitchen.yml for the omnibus install of chef.

**Comment by lamont-granquist** [comment]
^ does the `def action_install` vs `action :install` stuff fix you?

**Comment by lamont-granquist** [comment]
Simpler replication case:

```ruby
class MyProvider < Chef::Provider::LWRPBase
  use_inline_resources

  provides :test_test

  action :doit do
#  def action_doit
    file "/tmp/foobarbaz" do
      content "it works"
    end
  end
end

class MyResource < Chef::Resource::LWRPBase
  provides :test_test

  resource_name :test_test

  actions :doit
  default_action :doit
end
```

**Comment by lamont-granquist** [comment]
I validated that a normally written LWRP without use_inline_resources is not broken.

It also looks like it doesn't matter if you have use_inline_resources declared or not, inheriting from Chef::Provider:::LWRPBase and then using `def action_install` instead of `action :install` does not work in either case.

I'm still of the opinion that all the cookbooks need to get fixed because they were all broken to start with before 12.5.1 was released...

**Comment by ranjib** [comment]
well.. it used to work before.
i think we should make a call on whether we plan to support HWRP style, i.e. resource/provider that are written as standard ruby class, and if so what is the public API, so that we can use that to guide people whether a particular implementation will be supported or not (or buggy/uses internal api etc)

**Comment by lamont-granquist** [comment]
Again "work".  I've also fought with these problems in detail for a few months and made up *my* mind that the "HWRP Style" is wrong, fraught with peril, and is doing a massive disservice to less knowledgeable chef community members.  Whatever 'we' decide isn't gonna change *my* mind, I've seen enough of it.

I'm writing some foodcritic rules to enforce the use of use_inline_resources and enforce the use of the action DSL method over opening up your own action methods.

**Comment by coderanger** [comment]
@lamont-granquist As someone diametrically opposed, I hope those are going in your own ruleset ;-)

**Comment by lamont-granquist** [comment]
Nope, you can oppose, but we'll have to hammer it out at the meeting on thursday.

**Comment by lamont-granquist** [comment]
Also, it was @tas50's immediate request that we get FC rules in place to catch this.

**Comment by Sauraus** [comment]
Modifying the code to read ```action :install do``` fixes the problem on 12.5.1, although this fixes the issue I do not think that this change in chef-client can remain because it breaks existing cookbooks in a silent and nasty way.

IMHO this 'change' needs to be reverted and modify to throw a warnings before being implemented again in a subsequent MAJOR release of chef-client.

**Comment by lamont-granquist** [comment]
https://github.com/acrmp/foodcritic/pull/402

**Comment by josb** [comment]
Feel free to disagree, but this thread once again suggest to me that as a long-term direction the Chef Recipe DSL should be abandoned, and the community should move to the traditional object-oriented model to avoid these kinds of issues. The Recipe DSL may seem easy to use for beginners, but in my opinion it doesn't scale and actually hampers the growth of the tool.

**Comment by thommay** [comment]
@jkeiser can you take a look at this when you have some cycles?

**Comment by iroller** [comment]
I'm having a similar issue after upgrading from 12.4.0 to 12.5.1

https://github.com/hw-cookbooks/lxc/blob/master/providers/config.rb#L77

```
ArgumentError
  -------------
  wrong number of arguments (0 for 1)

  Cookbook Trace:
  ---------------
  /var/chef/cache/cookbooks/lxc/providers/config.rb:77:in `block (2 levels) in class_from_file'
  /var/chef/cache/cookbooks/lxc/providers/config.rb:75:in `block in class_from_file'
```

**Comment by someara** [comment]
ping @jkeiser

**Comment by jkeiser** [comment]
I should have a fix for this shortly.  The issue, as @lamont-granquist points out, is that when you put inline resources into your action, 12.5.1 now converges as part of `action :install`, while 12.4 did the converge in `run_action`. I'll restore the `run_action` converge, which should fix this.

@iroller that feels like a different issue, can you file a new bug?

**Comment by iroller** [comment]
@jkeiser sure, done https://github.com/chef/chef/issues/4224

**Comment by jkeiser** [comment]
And a note: it is *not* our intent to break any existing resources. Broadly, anything that worked in 12.4, must work in 12.5. Thanks for the bug :)

**Comment by someara** [comment]
<3






—

**Comment by jkeiser** [comment]
@Sauraus this should be fixed in 12.6 :) However, I'd still advise that you fix up the cookbooks using this pattern to use `action :create` instead of `def action_create`, because when you do `def`, the resources you add are added to the *end of the Chef run* instead of being executed inline. This causes two issues:

1. The actual inline resources won't run until *all other resources in your run list have converged*. This out of order stuff is going to seem weird to users, and will cause breakage in some cases.
2. Notifies and subscribes will not work on your resource, because it is never marked updated.

This will happen on all Chef versions--12.4 and below even--until it's fixed. Nonetheless, 12.6 should restore the 12.4 behavior, so it'll be the same level of bad instead of silently doing nothing, which sucks too :)
